### ISSUE TITLE
BUG: np.unwrap accumulates rounding errors

### ISSUE THREAD

**Issue by bmerry** [numpy/numpy#27609]
### Describe the issue:

When used with large values and on large arrays, the values towards the end of the array can have very large errors in phase. This is because the implementation calculates corrections to the differences between elements, then applies np.cumsum to it to determine the corrections to the elements. Since every term has rounding error, the tail of the cumsum has a large error.

This could be improved by calculating corrects in units of the period, which would be integral and hence only suffer from overflow when large enough not to fit into the floating-point type (2^53 for float64).

### Reproduce the code example:

```python
import numpy as np

tau = 2 * np.pi

def phase_error(x, y):
    return (x - y + np.pi) % tau - np.pi

x = np.random.uniform(-1e9, 1e9, size=64 * 1024 * 1024)
y = np.unwrap(x)
print("Max phase error for np.unwrap: ", np.max(np.abs(phase_error(x, y))))
# This gives more accurate results
z = np.unwrap(x / tau, period=1.0) * tau
print("Max phase error for adjusted np.unwrap: ", np.max(np.abs(phase_error(x, z))))
```

### Error message:

```shell
Max phase error for np.unwrap:  0.9471197530276747
Max phase error for adjusted np.unwrap:  3.597418789524909e-07
```

### Python and NumPy Versions:

2.1.2
3.12.3 (main, Sep 11 2024, 14:17:37) [GCC 13.2.0]

### Runtime Environment:

[{'numpy_version': '2.1.2',
  'python': '3.12.3 (main, Sep 11 2024, 14:17:37) [GCC 13.2.0]',
  'uname': uname_result(system='Linux', node='bender', release='6.8.0-45-generic', version='#45-Ubuntu SMP PREEMPT_DYNAMIC Fri Aug 30 12:02:04 UTC 2024', machine='x86_64')},
 {'simd_extensions': {'baseline': ['SSE', 'SSE2', 'SSE3'],
                      'found': ['SSSE3',
                                'SSE41',
                                'POPCNT',
                                'SSE42',
                                'AVX',
                                'F16C',
                                'FMA3',
                                'AVX2'],
                      'not_found': ['AVX512F',
                                    'AVX512CD',
                                    'AVX512_KNL',
                                    'AVX512_KNM',
                                    'AVX512_SKX',
                                    'AVX512_CLX',
                                    'AVX512_CNL',
                                    'AVX512_ICL']}},
 {'architecture': 'Haswell',
  'filepath': '/home/bmerry/.pyenv/versions/env/lib/python3.12/site-packages/numpy.libs/libscipy_openblas64_-ff651d7f.so',
  'internal_api': 'openblas',
  'num_threads': 32,
  'prefix': 'libscipy_openblas',
  'threading_layer': 'pthreads',
  'user_api': 'blas',
  'version': '0.3.27'}]

### Context for the issue:

To be honest, this is probably low priority. But I did discover this in real code which produced data which produced phases that were all expected to be roughly equal modulo multiples of 2π. I wanted to take the mean and standard deviation and used `np.unwrap` to condition the data so that these would make sense. While the errors were less extreme than in this example, they were enough to be cause real issues.

**Comment by sf2718** [comment]
Alternatively, one could rewrite this function such that it doesn't cumsum over non-integral arrays. I wrote such an implementation, and it does in fact get more accurate results (without the external division you did). The runtime is approximately the same as the original implementation (and even a little bit faster).
I time permits I will write a PR to commit this in soon.

```python
import numpy as np

def my_unwrap(p, discont=None, axis=-1, period=2*np.pi):
    p = np.asarray(p)
    dd = np.diff(p, axis=axis)
    dtype = np.result_type(dd, period)
    
    if np.issubdtype(dtype, np.integer):
        half_period, rem = np.divmod(period, 2)
        boundary_ambiguous = rem == 0
    else:
        half_period = period / 2
        boundary_ambiguous = True

    dd_mod_2period = np.mod(dd, 2 * period)
    dd_periods = np.round(dd / period).astype(int)

    if boundary_ambiguous:
        neg_corrections = (dd_mod_2period == half_period + period) & (dd > 0)
        pos_corrections = (dd_mod_2period == half_period) & (dd < 0)
        dd_periods[neg_corrections] -= 1
        dd_periods[pos_corrections] += 1

    if discont is not None and discont > half_period:
        dd_periods[np.abs(dd) < discont] = 0

    phase_shifts = np.cumsum(dd_periods, axis=axis)

    slices = [slice(None)] * p.ndim
    slices[axis] = slice(1, None)
    indices = tuple(slices)
    p_unwrapped = p.copy()
    p_unwrapped[indices] -= phase_shifts * period

    return p_unwrapped
```

**Comment by bmerry** [comment]
> Alternatively, one could rewrite this function such that it doesn't cumsum over non-integral arrays.

Yes, that's the kind of thing I had in mind - my external division was just a quick way to show that it's possible to do better than the current implementation.

By the way, the cast to `int` in your implementation may be problematic: if the corrections are larger than 2^63 periods (or 2^31 on 32-bit systems), you'll get overflows and *very* wrong answers, while if you keep things in floats then you'll get more graceful degradation (the phase will be garbage, but that's unavoidable since the inputs don't have the significant bits to capture the phase anyway).

**Comment by chrisjbillington** [comment]
This can be quite bad on relatively smaller amounts of data if you are using float32s

```python
import numpy as np
import matplotlib.pyplot as plt

phi = np.linspace(0, 2**16, 2**24, dtype=np.float32)
phi_roundtripped = np.unwrap(phi % (2 * np.pi))

plt.plot(phi_roundtripped - phi)
plt.xlabel('sample number')
plt.ylabel('unwrap error (rad)')
plt.show()
```

<img width="594" height="472" alt="Image" src="https://github.com/user-attachments/assets/de50ec92-fe56-429d-8003-7e33e61fc2c9" />
